content/6_Advanced/Treaps.mdx

```diff
@@ -326,6 +328,215 @@ With split and merge, how do we implement the additional operations?
   in lazy propogation). When splitting or merging, we push these tags downwards
   and perform the corresponding operation.
 
+### Implementation
+
+This code solves the above problem, but also provides a generalizable template.
+
+<LanguageSection>
+
+<CPPSection>
+
+```cpp
+// https://judge.yosupo.jp/problem/dynamic_sequence_range_affine_range_sum
+#include <bits/stdc++.h>
+#define int long long
+using namespace std;
+
+// BeginCodeSnip{ModInt class}
+// operations under mod
+struct mint;
+using P = pair<mint, mint>;
+struct mint {
+	const static int M = 998244353;
+	int v = 0;
+	mint() {}
+	mint(int v) { this->v = (v % M + M) % M; }
+	mint operator+(const mint &o) const { return v + o.v; }
+	mint &operator+=(const mint &o) {
+		v = (v + o.v) % M;
+		return *this;
+	}
+	mint operator*(const mint &o) const { return v * o.v; }
+	mint operator-(const mint &o) const { return v - o.v; }
+	mint &operator-=(const mint &o) {
+		mint t = v - o.v;
+		v = t.v;
+		return *this;
+	}
+	mint operator^(int y) const {
+		mint r = 1, x = v;
+		for (y <<= 1; y >>= 1; x = x * x)
+			if (y & 1) r = r * x;
+		return r;
+	}
+	mint inv() const {
+		assert(v);
+		return *this ^ M - 2;
+	}
+	friend istream &operator>>(istream &s, mint &v) {
+		s >> v.v;
+		return s;
+	}
+	friend ostream &operator<<(ostream &s, const mint &v) {
+		s << v.v;
+		return s;
+	}
+	mint operator/(mint o) { return *this * o.inv(); }
+};
+// EndCodeSnip
+
+struct F {  // linear function wx + b
+	mint w = 1, b = 0;
+	mint operator()(mint x) { return w * x + b; }
+	F operator()(F f) { return F{w * f.w, w * f.b + b}; }
+	operator bool() const {
+		return w.v != 1 || b.v != 0;
+	}  // false if this is identity function
+};
+
+mt19937_64 rng(chrono::steady_clock::now().time_since_epoch().count());
+
+struct Node {
+	Node(mint val) : val(val), sum(val), weight(rng()), size(1) {}
+	mint val, sum;  // val -> a[i], sum = sum of all a[i] in subtree
+	int weight, size;
+	bool rev = false;  // whether this range is reversed
+	Node *l = NULL, *r = NULL;
+	F f;  // lazy affine tag
+};
+
+int size(Node *treap) { return treap ? treap->size : 0; }
+mint sum(Node *treap) { return treap ? treap->sum : 0; }
+void push(Node *treap) {
```

Comment:
doc comments would be cool

Reply:
also why not make these methods

Reply:
cuz `treap` could be nullptr

Reply:
pretty sure it should be the caller's responsibility to make sure that's the case(?)

Reply:
idk most treap implementation's ive seen have it this way so you dont have to check treap->l and treap->r every single time

Reply:
> doc comments would be cool

the names feel pretty self-explanatory though